### ISSUE TITLE
pointer solution not used when parameter is defined as omitempty, 0 value is not submitted

### ISSUE THREAD

**Issue by iafilius** [go-swagger/go-swagger#2166]
## Problem statement

Hello,

small introduction:
As described in #2156, i'm working on generating and testing client library for a swagger specified on:
https://vdc-download.vmware.com/vmwb-repository/dcr-public/6faa975e-68ee-4c6c-944a-5e9e664b2475/50c61a14-cc14-4adb-b936-b5411ccdd1a5/swagger-3.3.2-dist.json.
Which is latest 3.3.2 version from https://code.vmware.com/apis/784/velocloud-sdwan-vco-api

Fixed all items for duplicated definitions, and a few other "cornercases" which are addressed in #2161 for follow-up.

The swagger version derived from that, with all required fixes are at:
[Manual fixed swagger](https://iafilius.home.xs4all.nl/swagger-3.3.2-dist-manualfixes7.json)

Client generated with:
```
../go-swagger/cmd/swagger/swagger generate client -f ./swagger-3.3.2-dist-manualfixes7.json ...
```

It looks like overlap with issue #1707 which is open for more then a year. (but not 100% sure)

While starting with the most simple API call in library (apart from logging in), i need to give the 'EdgeEdgeProvisionBody' struct as API call parameter.

As i'm lookin into go-swagger as an option to generate the client library (opposed to maven/swagger UI), with special focus on a few known issues like:
- not submitting 'non defined and optional(omitempty) types' (structs, bool's, int's, strings etc), even when a possible default value is zero (by design)
- submitting 'defined and optional (omitempty) typed' (structs, bool's, int's, strings etc), even when value is zero.

Using pointer's to (nested) types (struct, string, bool, int ...) is a good working solution path, and also practiced in go-swagger.
But it looks like the produced code contains an error.
For example, the first struct i noticed:
```
/*EdgeEdgeProvisionBody edge_edge_provision
swagger:model EdgeEdgeProvisionBody
*/
type EdgeEdgeProvisionBody struct {

	// challenge password
	ChallengePassword string `json:"challengePassword,omitempty"`

	// configuration Id
	// Required: true
	ConfigurationID *int64 `json:"configurationId"`

	// description
	Description string `json:"description,omitempty"`

	// enterprise Id
	EnterpriseID int64 `json:"enterpriseId,omitempty"`

	// generate certificate
	GenerateCertificate bool `json:"generateCertificate,omitempty"`

	// ha enabled
	HaEnabled bool `json:"haEnabled,omitempty"`

	// model number
	// Required: true
	// Enum: [edge500 edge5X0 edge510 edge6X0 edge840 edge1000 edge1000qat edge3X00 virtual]
	ModelNumber *string `json:"modelNumber"`

	// name
	Name string `json:"name,omitempty"`

	// private key password
	PrivateKeyPassword string `json:"privateKeyPassword,omitempty"`

	// serial number
	SerialNumber string `json:"serialNumber,omitempty"`

	// site
	Site *models.Site `json:"site,omitempty"`

	// subject c n
	SubjectCN string `json:"subjectCN,omitempty"`

	// subject o
	SubjectO string `json:"subjectO,omitempty"`

	// subject o u
	SubjectOU string `json:"subjectOU,omitempty"`
}
```

Solution for nested structs looks OK:
```
	// site
	Site *models.Site `json:"site,omitempty"`
```

Solution for Mandatory values has the pointer solution, so it will submit the parameter but with null value, like:
'''
"configurationId":null,
'''
This is perfectly fine.

But the types which are defined defined as optional (omitempty) are not defined as pointers. A typical sample:
```
	// ha enabled
	HaEnabled bool `json:"haEnabled,omitempty"`
```
So in this case, the boolean only wil be submitted when having the true value, and false value is always ignored. (empty/not defined != false)

The same with an int:
```
	// enterprise Id
	EnterpriseID int64 `json:"enterpriseId,omitempty"`

```
So a valid value of '0' always will be ignored.
This last case (omitempty without pointer solution) should be changed to a using a pointer as well.
Removing the 'omitempty' does not solve it, it introduces the issue that value always is send, even when not defiend (as null, when using pointer).
In this case the API server is quite unpredictable when submitting parameters is is not expecting, introducing lots of undesired effects.

Regards,

Arjan

## Swagger specification
2.0
## Steps to reproduce

See above description, 

## Environment
swagger version: 
../go-swagger/cmd/swagger/swagger version
dev

git log:
commit 5917c9a6d36b7b356fab7c0378f36468c6de206d 
Author: autumn31 <autumn.forever31@gmail.com>
Date:   Tue Dec 31 19:27:58 2019 +0800
  
$ go version
go version go1.12.10 linux/amd64

OS:   
"Ubuntu 19.10"

**Comment by fredbi** [comment]
@iafilius yes this is a design choice to avoid putting pointers everywhere. Whenever you need to explicitly know about the presence of a zero value (such a `0`, `false`, `""`), you need to override the generation by adding a `x-nullable: true`  extension.

**Comment by iafilius** [comment]
Thanks Fredbi,
Thanks for your directions
Added the x-nulable to all parameters in one API body object which are "omitempty" (apart the nested struct Site object, which is omitempty, and is already a pointer: 
```
                "description": {
                  "type": "string",
                  "x-nullable": true
                },
                "site": {
                  "$ref": "#/definitions/site"
                },
                "haEnabled": {
                  "type": "boolean",
                  "x-nullable": true
                },
                "generateCertificate": {
                  "type": "boolean",
                  "x-nullable": true
                },
                "subjectCN": {
                  "type": "string",
                  "x-nullable": true
                },
                "subjectO": {
                  "type": "string",
                  "x-nullable": true
                },
                "subjectOU": {
                  "type": "string",
                  "x-nullable": true
                },
                "challengePassword": {
                  "type": "string",
                  "x-nullable": true
                },
                "privateKeyPassword": {
                  "type": "string",
                  "x-nullable": true
                }
              },
```
Re-generated the library and that works great, see result:
```
/*EdgeEdgeProvisionBody edge_edge_provision
swagger:model EdgeEdgeProvisionBody
*/
type EdgeEdgeProvisionBody struct {

	// challenge password
	ChallengePassword *string `json:"challengePassword,omitempty"`

	// configuration Id
	// Required: true
	ConfigurationID *int64 `json:"configurationId"`

	// description
	Description *string `json:"description,omitempty"`

	// enterprise Id
	EnterpriseID *int64 `json:"enterpriseId,omitempty"`

	// generate certificate
	GenerateCertificate *bool `json:"generateCertificate,omitempty"`

	// ha enabled
	HaEnabled *bool `json:"haEnabled,omitempty"`

	// model number
	// Required: true
	// Enum: [edge500 edge5X0 edge510 edge6X0 edge840 edge1000 edge1000qat edge3X00 virtual]
	ModelNumber *string `json:"modelNumber"`

	// name
	Name *string `json:"name,omitempty"`

	// private key password
	PrivateKeyPassword *string `json:"privateKeyPassword,omitempty"`

	// serial number
	SerialNumber *string `json:"serialNumber,omitempty"`

	// site
	Site *models.Site `json:"site,omitempty"`

	// subject c n
	SubjectCN *string `json:"subjectCN,omitempty"`

	// subject o
	SubjectO *string `json:"subjectO,omitempty"`

	// subject o u
	SubjectOU *string `json:"subjectOU,omitempty"`
}
```
And with that the desired behaviour is send on API level, when not defined, value is not submitted, if set to any value (including zero) it is submitted. Happy with that!
Thanks you!

The drawback is with a 764K(bytes) 3th party swagger file, with maybe thousands of parameters with omitempty, the spec file seems not fully stabilized and so i expect a new version on regular base.
That would generate a lot of manual /recurring work.
In trying to prevent that, I looked into setting (by patching the code) the default for x-nullable for those parameters which do not yet use the pointer construction by default. I'm affraid i could not find a clear point (yet) in the code to where that possible could be done.

Could you give a hint please, and/or may i request a 'feature enhancement' for it?

Thanks for considering.

Arjan

**Comment by fredbi** [comment]
Yes I also have met the same issue with big specs... How would suggest this to be working? IMHO, we may add some CLI options, but that could only be a global setting.
On the other hand, we could add some features to the spec manipulation kit (go-openapi/analysis) which current can bring fixes, flatten, diff and merge specs. TL,DR we know there is something to be improved here, but not so sure about how best to proceed.

**Comment by iafilius** [comment]
Hello Fredbi,

thanks for considering, i'd made a list based on arguments i see, in combination with estimated/indication of efford it could take.

# In short for a workable solution (in order)

All solutions i talk on, are considered global, i do not see any use for enabeling the nullable, and omitempty capability (including support the valid value of 0) just for a subgroup.

- solution/support in go-swagger for both nullable, and ommitempty out of the box (depending when value is not set, so value of 0 is allowed and != 'not set') ). This implies changing default in current go-swagger code (at least for those is currently isn't yet).
    - drawback rethink previous made design decision(s)
    - possible long time to implement (due discussions)
    - somewhat more complexity using client generated library
    - additional internal complexity/memory managent using pointers/allocations
    - solution is available by whole community, but changing default would force current users to slighty rewrite their code for using the generated library.
- I would be realy happy 'just' with a CLI/global switch option in go-swagger.
    - looks like more (or less) easy to implement
    - hopefully would not need too much time/discussion.
    - generates only the internal complexity/memory management when requested by CLI option.
    - solution is available by whole community, but changing default by CLI would only those users who need/want it, but do need to slighty rewrite their code for using the generated library with extra CLI switch
- manual applying a small patch (to go-swagger code) each time i need it. (but i'm not sure where and what to patch yet, even not sure it would be small)
    - Drawback on our side having to find out where and what to patch, and need it supported for futre go-swagger revisions as well
    - not having this capability out of the box available in go-swagger (not by default accessible by community)
- Not completely sure regarding  spec manipulation kit (go-openapi/analysis), but an offline tool adding the x-nullable option on the right places.
Those are the workable solutions I see.
    - would just need a small extra processing step (converting swagger spec)
    - unsure on time/complexity
    - solution accessible by community/others
- at last, manual patching all types in the swagger specfile by each user themself. 
    - no community benefit
    - maybe we can use ansible to help us with that.
    - might be fastest to start some further integration tests to fully discover the go-swagger generated client library.

# Preferred solution (long)

Note: Not trying to redo previous design decisions and discussions, but my opinion/view on the matter.

## Prefered capabilities

Preferred having both the capabilities by default out of the box (as i think that is the intended way), on global level:

- nullable ('variabele: null' in json) when value is not set. If variable set to any value including '0' (or empty string and ...), submit 'variable: value'.
- not submitting empty key/struct/.. when 'omitempty' and where value is not set.  If variable set to any value including '0' (or empty string and ...), submit 'variable: value'.

## Regarding implementations/technology choices

### Pointers

With the pointer solution you bring basically an extra meta data to fascillitate the 2/both mentionned capabilities, making the distinction (based on pointer value) if that variable is set/used, or not.

The pointer solutions works perfectly for languages which do no know the difference between having a variabele set to zero or having still the default of zero, but at a cost forcing all paramters be-ing pointer (adding complexity on client side code, and library generation side)

### Other

I guess using pointer is not the only solution which could have been used, you 'just' need a way to hold that extra meta data (is the variabele set, or not)
Altough i think this might possible, it would be more complex, and perhaps less straight forward as using the pointer for the meta data (variabele/key is set or not)
A basic sample of that idea
```
type SwaggerInt struct {
    bool used       // Holds the meta data is value is used
    RealInt int
}
```
and possible put it into a calls, and define operations on it.
Anyway the pointer sulution will bring both capabilities, and is currently already available (only not default out of the box, and a bit ackward to enable for all paramters)



Do you have enough feedback on this?

Regards,

**Comment by casualjim** [comment]
Hi I like option 2 and 4 from your proposals. I imagine option 4 is the easiest to do because you just define a set of rules.  

You can find a very old branch that basically does option 1: https://github.com/go-swagger/go-swagger/pull/557
Some discussion here: https://github.com/go-swagger/go-swagger/issues/476

Feel free to submit a PR that does either one of these (2 or 4) :)

**Comment by iafilius** [comment]
for now just patched the (huge) swagger spec to unblock me for further testing.
Not sure if i should close this issue, or just keep it as ‘enhancement’ open?

Regards

**Comment by joelterry** [comment]
I just wanted to throw in my lot behind the users that have opened issues on the topic of omitempty for optional values. Number/boolean pointers are weird, but match up better with optional JSON values. I'm relatively unfamiliar with Swagger, but it seems like being language agnostic should be a greater priority than avoiding Go-specific weirdness. Meaning, the default configuration should try to accomodate the use case of cross-language communication, and then a cli setting could enable omitted 0/false if the user knows the receiving end will react accordingly.

I'm passionate about this issue because I was just hit with it: I had to edit a third party spec to fix the bug, and now have to keep track of how I changed it and stay on alert (when ideally the spec would have been a source of truth). It seems like it would be a common footgun for others as well. That being said, this is a complex project, so I understand if this type of change is out of the question at this point. And maybe this opinion is less popular than I imagine 🤷
